source/_posts/vue-design-patterns.md

```diff
@@ -0,0 +1,988 @@
+      }
+    
+      remove(key: string) {
+        this.cookies.remove(key);
+      }
+    }    
+```
+
+If one day we will need to replace the 3rd party library or add more features, it will be fast, and many parts of your code won't be affected by this change, as we can see in the following schematic image.
+
+{% img "adapter-pattern.jpg" "Adapter pattern schema" "lazy" %}
+
+Let's think about another situation: You have an extensive web application with dozens of elements from some UI material library/framework (like Vuetify). Tomorrow, the business requirements change to adopt Quasar because of the long-term support. You have picked some elements from Vuetify, and now it's time to change it 😨. But this scenario happens and could be acceptable, but as time goes on, technologies improve, and again a change is needed. What tedious work. The solution? Create your adapter input😎.
+
+
+## Container/Presentational Pattern
+Maybe you are using this design pattern right now, it's simple, and that's the point. Vue and other modern frameworks have two parts where you can put the code, the presentational layer and the business logic layer. Those are known as concerns, and generally isn't convenient to mix them up, but it doesn't mean that one can't use the other, and this is where Container or Presentational Pattern lands.
+
+It's time to imagine another scenario: You need a list of To-Do items. The easy way is to create a Vue SFC (Single File Component) to show all of them, something like this.
+
+```ts
+    // File: TodoList.vue
+    
+    <template>
+      <div class="todo" v-for="(todo, index) in todos.slice(0, 10)" :key="index">
+        <span :class="['todo__title', todo.completed && 'todo__title--completed']">
+          {{ todo.title }}
+        </span>
+        <span class="todo__completed">
+          {{ todo.completed ? "👍" : "👎" }}
+        </span>
+      </div>
+    </template>
+    
+    <script lang="ts">
+    import { defineComponent, ref, type Ref } from "vue";
+    
+    export interface Todo {
+      title: string;
+      completed: boolean;
+    }
+    
+    export default defineComponent({
+      setup() {
+        const todos: Ref<Todo[]> = ref([]);
+        const getData = async () => {
+          const data = await fetch("https://jsonplaceholder.typicode.com/todos");
+          todos.value = await data.json();
+        };
+    
+        getData();
+        return {
+          todos,
+        };
+      },
+    });
+    </script>
+    
+    <style scoped>
+    .todo {
+      margin: 0.5rem;
+      border-radius: 0.5rem;
+      background-color: rgb(224, 224, 224);
+      color: black;
+      padding: 0.5rem;
+    }
+    
+    .todo:hover {
+      scale: 1.1;
+    }
+    
+    .todo__title {
+      color: rgb(216, 55, 55);
+    }
+    .todo__title--completed {
+      text-decoration: line-through;
+      color: rgb(106, 163, 21);
+    }
+    
+    .todo__completed {
+      float: right;
+    }
+    </style>    
```

Comment:
I have a feeling that the CSS styles are not relevant to this article. What do you think?
The style is **not the core** of the main idea (design patterns in TS code). They make the code examples very bloated...

Reply:
I've removed the style tag in all examples.